### PR TITLE
Fix the order while sorting the operators

### LINKED ISSUE #16316: test_analyzer_transformer random fails on compare
http://ci.paddlepaddle.org/viewLog.html?buildId=72226&tab=buildLog&buildTypeId=Paddle_PrCiPython35Night&logTab=tree&filter=all&_focus=13745
```
[23:57:29][ RUN      ] Analyzer_Transformer.compare
[23:57:29]I0319 23:55:10.920542 17062 analyzer_transformer_tester.cc:176] The number of samples to be test: 8
[23:57:29]I0319 23:55:10.922397 17062 tester_helper.h:65] AnalysisConfig {
[23:57:29]  NativeConfig {
[23:57:29]    PaddlePredictor::Config {
[23:57:29]      model_dir: 
[23:57:29]    }
[23:57:29]    use_gpu: 0
[23:57:29]    device: 0
[23:57:29]    fraction_of_gpu_memory: 0.00219227
[23:57:29]    specify_input_name: 1
[23:57:29]    cpu_num_threads: 1
[23:57:29]  }
[23:57:29]  prog_file: /root/.cache/inference_demo/transformer/model/model
[23:57:29]  param_file: /root/.cache/inference_demo/transformer/model/params
[23:57:29]  enable_ir_optim: 1
[23:57:29]  enable_ir_optim: 1
[23:57:29]  use_feed_fetch_ops: 1
[23:57:29]  use_tensorrt: 0
[23:57:29]  use_mkldnn: 0
[23:57:29]}
[23:57:29]W0319 23:55:10.925616 17062 init.cc:153] AVX is available, Please re-compile on local machine
[23:57:29]I0319 23:55:11.322595 17062 tester_helper.h:278] Running thread 0, warm up run...
[23:57:29]I0319 23:55:44.447240 17062 helper.h:270] ====== batch_size: 8, repeat: 1, threads: 1, thread id: 0, latency: 33124.5ms, fps: 0.0301891 ======
[23:57:29]I0319 23:55:44.447353 17062 tester_helper.h:301] Thread 0 run 1 times...
[23:57:29]I0319 23:56:17.681612 17062 helper.h:270] ====== batch_size: 8, repeat: 1, threads: 1, thread id: 0, latency: 33234.2ms, fps: 0.0300895 ======
[23:57:29]W0319 23:56:17.688809 17062 init.cc:153] AVX is available, Please re-compile on local machine
[23:57:29]--- Running analysis [ir_graph_build_pass]
[23:57:29]--- Running analysis [ir_analysis_pass]
[23:57:29]--- Running IR pass [infer_clean_graph_pass]
[23:57:29]--- Running IR pass [attention_lstm_fuse_pass]
[23:57:29]--- Running IR pass [seqpool_concat_fuse_pass]
[23:57:29]--- Running IR pass [seqconv_eltadd_relu_fuse_pass]
[23:57:29]--- Running IR pass [fc_lstm_fuse_pass]
[23:57:29]--- Running IR pass [mul_lstm_fuse_pass]
[23:57:29]--- Running IR pass [fc_gru_fuse_pass]
[23:57:29]--- Running IR pass [mul_gru_fuse_pass]
[23:57:29]--- Running IR pass [seq_concat_fc_fuse_pass]
[23:57:29]--- Running IR pass [fc_fuse_pass]
[23:57:29]---  detected 12 subgraphs
[23:57:29]--- Running IR pass [repeated_fc_relu_fuse_pass]
[23:57:29]--- Running IR pass [squared_mat_sub_fuse_pass]
[23:57:29]--- Running IR pass [conv_bn_fuse_pass]
[23:57:29]--- Running IR pass [conv_eltwiseadd_bn_fuse_pass]
[23:57:29]--- Running IR pass [is_test_pass]
[23:57:29]--- Running IR pass [identity_scale_op_clean_pass]
[23:57:29]--- Running analysis [ir_params_sync_among_devices_pass]
[23:57:29]--- Running analysis [ir_graph_to_program_pass]
[23:57:29]I0319 23:56:24.455240 17062 analysis_predictor.cc:396] == optimize end ==
[23:57:29]I0319 23:56:24.456599 17062 tester_helper.h:278] Running thread 0, warm up run...
[23:57:29]W0319 23:56:24.456904 17062 naive_executor.cc:43] The NaiveExecutor can not work properly if the cmake flag ON_INFER is not set.
[23:57:29]W0319 23:56:24.456938 17062 naive_executor.cc:45] Unlike the training phase, all the scopes and variables will be reused to save the allocation overhead.
[23:57:29]W0319 23:56:24.456956 17062 naive_executor.cc:48] Please re-compile the inference library by setting the cmake flag ON_INFER=ON if you are running Paddle Inference
[23:57:29]I0319 23:56:56.635799 17062 helper.h:270] ====== batch_size: 8, repeat: 1, threads: 1, thread id: 0, latency: 32179.1ms, fps: 0.031076 ======
[23:57:29]I0319 23:56:56.635892 17062 tester_helper.h:301] Thread 0 run 1 times...
[23:57:29]W0319 23:56:56.636173 17062 naive_executor.cc:43] The NaiveExecutor can not work properly if the cmake flag ON_INFER is not set.
[23:57:29]W0319 23:56:56.636209 17062 naive_executor.cc:45] Unlike the training phase, all the scopes and variables will be reused to save the allocation overhead.
[23:57:29]W0319 23:56:56.636229 17062 naive_executor.cc:48] Please re-compile the inference library by setting the cmake flag ON_INFER=ON if you are running Paddle Inference
[23:57:29]I0319 23:57:28.508708 17062 helper.h:270] ====== batch_size: 8, repeat: 1, threads: 1, thread id: 0, latency: 31872.8ms, fps: 0.0313748 ======
[23:57:29]/paddle/paddle/fluid/inference/tests/api/tester_helper.h:82: Failure
[23:57:29]      Expected: size
[23:57:29]      Which is: 645
[23:57:29]To be equal to: ref_size
[23:57:29]      Which is: 666
[23:57:29]/paddle/paddle/fluid/inference/tests/api/tester_helper.h:89: Failure
[23:57:29]      Expected: pdata_ref[j]
[23:57:29]      Which is: 6444
[23:57:29]To be equal to: pdata[j]
[23:57:29]      Which is: 23118
[23:57:29]/paddle/paddle/fluid/inference/tests/api/tester_helper.h:89: Failure
[23:57:29]      Expected: pdata_ref[j]
[23:57:29]      Which is: 10
[23:57:29]To be equal to: pdata[j]
[23:57:29]      Which is: 855
```

### LINKED ISSUE #16473: test_analyzer_small_dam random fails on accuracy
```
[02:50:55]	I0327 02:50:55.226866 88684 helper.h:270] ====== batch_size: 1, repeat: 1, threads: 1, thread id: 0, latency: 34.1772ms, fps: 29.2592 ======
[02:50:55]	I0327 02:50:55.226940 88684 tester_helper.h:301] Thread 0 run 1 times...
[02:50:55]	I0327 02:50:55.258460 88684 helper.h:270] ====== batch_size: 1, repeat: 1, threads: 1, thread id: 0, latency: 31.4899ms, fps: 31.7562 ======
[02:50:55]	F0327 02:50:55.268436 88684 tester_helper.h:97] Check failed: std::abs(pdata_ref[j] - pdata[j]) <= FLAGS_accuracy (0.201252 vs. 0.001) 
```
http://ci.paddlepaddle.org/viewLog.html?buildId=75853&tab=buildLog&buildTypeId=Paddle_PrCi&logTab=tree&filter=all&_focus=25726
Is this issue related with #16316?

### LINKED ISSUE #16586: random failure in processing seqpool_concat_fuse_pass when doing stress test for test_analyzer_small_dam
**System information**
-PaddlePaddle version: develop branch
-CPU: CPUMKL ON/OFF 
-GPU: No
-OS Platform: Linux
-Python version:  N/A
-Cmake orders
-C++version.txt
-API information
**To Reproduce**
Run "ctest -R test_analyzer_small_dam -V" for more than 5000 tests

**Describe your current behavior**
Random failure in seqpool_concat_fuse_pass processing in both builds with cmake option -DWITH_MKL=ON/OFF

**Code to reproduce the issue**

**Other info / logs**
```
116: Test command: /home/leozhao/Paddle/build/paddle/fluid/inference/tests/api/test_analyzer_small_dam "--inference_model_dir=/home/leozha
o/Paddle/build/third_party/inference_demo/word2vec/word2vec.inference.model" "--infer_model=/home/leozhao/Paddle/build/third_party/inferen
ce_demo/small_dam/model" "--infer_data=/home/leozhao/Paddle/build/third_party/inference_demo/small_dam/data.txt" "--max_turn_num=1"
116: Environment variables:
116:  FLAGS_cudnn_deterministic=true
116: Test timeout computed to be: 600
116: [==========] Running 8 tests from 1 test case.
116: [----------] Global test environment set-up.
116: [----------] 8 tests from Analyzer_dam
116: [ RUN      ] Analyzer_dam.profile
116: WARNING: Logging before InitGoogleLogging() is written to STDERR
116: I0331 21:58:43.398690 27293 analyzer_dam_tester.cc:178] The number of samples to be test: 1
116: I0331 21:58:43.398905 27293 tester_helper.h:65] AnalysisConfig {
116:   NativeConfig {
116:     PaddlePredictor::Config {
116:       model_dir:
116:     }
116:     use_gpu: 0
116:     device: 0
116:     fraction_of_gpu_memory: 0
116:     specify_input_name: 1
116:     cpu_num_threads: 1
116:   }
116:   prog_file: /home/leozhao/Paddle/build/third_party/inference_demo/small_dam/model/__model__
116:   param_file: /home/leozhao/Paddle/build/third_party/inference_demo/small_dam/model/param
116:   enable_ir_optim: 1
116:   enable_ir_optim: 1
116:   use_feed_fetch_ops: 1
116:   use_tensorrt: 0
116:   use_mkldnn: 0
116: }
116: ^[[1m^[[35m--- Running analysis [ir_graph_build_pass]^[[0m
116: ^[[1m^[[35m--- Running analysis [ir_analysis_pass]^[[0m
116: ^[[32m--- Running IR pass [infer_clean_graph_pass]^[[0m
116: ^[[32m--- Running IR pass [attention_lstm_fuse_pass]^[[0m
116: ^[[32m--- Running IR pass [seqpool_concat_fuse_pass]^[[0m
1/1 Test #116: test_analyzer_small_dam ..........***Exception: Other  1.25 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   1.28 sec
```

### LINKED ISSUE #16688: The order issue of graph_helper's unordered_set
在复现[ISSUE#16316](https://github.com/PaddlePaddle/Paddle/issues/16316)时候发现，graph_helper里使用了unordered_set来保存节点，但是发现unordered_set元素的插入顺序和再重新读出来的顺序不一致。 
    1）PaddlePaddle版本: 76b49f02ee642f83f4965a86ec08dad9219fa91b
    2）CPU: No AVX / No MKLML
    3）系统环境: CI 节点 yq01-gpu-255-134-14-04  
    4）尝试解决思路:
            将排序部分移至graph转program之前或是将unordered_set改为vector。
    5）Log Patch:
```
        diff --git a/paddle/fluid/framework/ir/graph_helper.cc b/paddle/fluid/framework/ir/graph_helper.cc
        index 28a37f3..b54e396 100644
        --- a/paddle/fluid/framework/ir/graph_helper.cc
        +++ b/paddle/fluid/framework/ir/graph_helper.cc
        @@ -143,8 +143,20 @@ std::map<ir::Node *, std::unordered_set<ir::Node *>>                       BuildOperationAdjList(
             std::sort(nodes.begin(), nodes.end(), [](ir::Node *node1, ir::Node *node2) {
               return node1->id() > node2->id();
     });
        +    std::cout << "vector ";
        +    for (auto it = nodes.begin(); it != nodes.end(); it++) {
        +      std::cout << "node id " << (*it)->id() << ", name " << (*it)->Name()
        +                << ", ";
        +    }
        +    std::cout << std::endl;
             adj_list[n].insert(std::make_move_iterator(nodes.begin()),
                                std::make_move_iterator(nodes.end()));
        +    std::cout << "unordered_set ";
        +    for (auto it = adj_list[n].begin(); it != adj_list[n].end(); it++) {
        +      std::cout << "node id " << (*it)->id() << ", name " << (*it)->Name()
        +                << ", ";
        +    }
        +    std::cout << std::endl;
           }
           return adj_list;
         }
```
    6）Log:
```
        [08:42:22] :	 [Step 1/1] 131: **vector node id 717, name assign, node id 715, name assign, node id 695, name assign,** node id 693, name assign, node id 673, name assign, node id 671, name assign, node id 651, name assign, node id 649, name assign, node id 629, name assign, node id 627, name assign, node id 607, name assign, node id 605, name assign, node id 585, name fill_constant_batch_size_like, node id 583, name fill_constant_batch_size_like, node id 581, name fill_constant_batch_size_like, node id 579, name fill_constant_batch_size_like, node id 577, name fill_constant_batch_size_like, node id 575, name fill_constant_batch_size_like, node id 573, name fill_constant_batch_size_like, node id 571, name fill_constant_batch_size_like, node id 569, name fill_constant_batch_size_like, node id 567, name fill_constant_batch_size_like, node id 565, name fill_constant_batch_size_like, node id 563, name fill_constant_batch_size_like, node id 561, name write_to_array, node id 559, name write_to_array, node id 554, name less_than, node id 552, name fill_constant, node id 550, name fill_constant, node id 3, name feed, node id 0, name feed, 
        [08:42:22] :	 [Step 1/1] 131: unordered_set node id 0, name feed, node id 3, name feed, node id 550, name fill_constant, node id 552, name fill_constant, node id 559, name write_to_array, node id 554, name less_than, node id 563, name fill_constant_batch_size_like, node id 565, name fill_constant_batch_size_like, node id 567, name fill_constant_batch_size_like, node id 569, name fill_constant_batch_size_like, node id 571, name fill_constant_batch_size_like, node id 573, name fill_constant_batch_size_like, node id 575, name fill_constant_batch_size_like, node id 577, name fill_constant_batch_size_like, node id 605, name assign, node id 607, name assign, node id 579, name fill_constant_batch_size_like, node id 581, name fill_constant_batch_size_like, node id 583, name fill_constant_batch_size_like, node id 585, name fill_constant_batch_size_like, node id 649, name assign, node id 651, name assign, node id 671, name assign, node id 673, name assign, node id 627, name assign, node id 693, name assign, node id 629, name assign, node id 695, **name assign, node id 715, name assign, node id 561, name write_to_array, node id 717, name assign**,
```       
    7）相关链接:
        https://stackoverflow.com/questions/41334952/how-does-unordered-set-determine-the-inserting-order-in-c

### PR DESCRIPTION
**Pull request by yihuaxu**
As [ISSUE#16688](https://github.com/PaddlePaddle/Paddle/issues/16688) description, it was found the order's random will result in the difference when do the compare item of unit test. Now to use set to replace unordered_set can be resolve this issue.

 resolve #16688 
 resolve #16316 
 resolve #16473 
 resolve #16586

### CODE REVIEW

**Comment by yihuaxu** [comment]
#16316

**Comment by yihuaxu** [comment]
#16316

**Comment by yihuaxu** [comment]
#16316

**Comment by yihuaxu** [comment]
Issue #16316

**Comment by yihuaxu** [comment]
start a review

**Comment by luotao1** [comment]
LGTM

File: paddle/fluid/framework/ir/graph_helper.cc [hunk 1]
```diff
@@ -32,7 +32,7 @@ namespace framework {
 namespace ir {
 namespace {
 void SortHelper(
-    const std::map<ir::Node *, std::unordered_set<ir::Node *>> &adj_list,
+    const std::map<ir::Node *, std::vector<ir::Node *>, ir::NodeComp> &adj_list,
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/graph_helper.cc]: try just change unordered_set to `std::set` ? 

the `set` is a binary search tree and should work like a sorted vector in this scenario.

Comment (yihuaxu) [reply paddle/fluid/framework/ir/graph_helper.cc]: Done, it should be OK. please wait CI's result.

Comment (yihuaxu) [reply paddle/fluid/framework/ir/graph_helper.cc]: Just now all CI testing are passed. Please help us review it again.

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**